subsys/fs/nvs/nvs.c

```diff
@@ -162,214 +503,215 @@ void _nvs_entry_sector_advance(struct nvs_fs *fs)
+			if (rc) {
+				return rc;
+			}
 		}
-		adv_len = _nvs_entry_len_in_flash(fs, head.len);
-		_nvs_addr_advance(fs, &entry->data_addr, adv_len);
 	}
+	return 0;
 }
 
-
-int nvs_get_last_entry(struct nvs_fs *fs, struct nvs_entry *entry)
+int nvs_reinit(struct nvs_fs *fs)
 {
-	int rc;
-	struct nvs_entry latest;
-	struct _nvs_data_hdr head;
-	off_t hdr_addr;
-	u16_t adv_len;
+	int rc, i;
+	u8_t sectors_in_use = 0, invalid_sectors = 0;
+	u16_t id, entry_sector_id = 0;
+	size_t len1, len2, entry_len;
+	off_t addr;
+	struct _nvs_sector_hdr sector_hdr;
+
+
+	k_mutex_lock(&fs->nvs_lock, K_FOREVER);
+
+	/* count the number of valid sectors_used */
+	for (i = 0; i < fs->sector_count; i++) {
+		addr = i * fs->sector_size;
+		rc = _nvs_sector_is_used(fs, addr);
+		if (rc < 0) {
+			goto err;
+		}
+		if (rc) {
+			rc = _nvs_flash_rd_len_id_start(fs, addr, &id, &len1);
+			if (rc) {
+				goto err;
+			}
+			if (id != NVS_ID_SECTOR_START) {
+				invalid_sectors++;
+				continue;
+			}
+			rc = _nvs_flash_read_data(fs, addr, &sector_hdr,
+					sizeof(sector_hdr));
+			if (rc) {
+				goto err;
+			}
+			if (sector_hdr.fd_magic != fs->magic) {
+				invalid_sectors++;
+				continue;
+			}
+			sectors_in_use++;
+			if (sectors_in_use == 1) {
+				entry_sector_id = sector_hdr.fd_id;
+				fs->entry_sector = i;
+				fs->sector_id = entry_sector_id;
+				continue;
+			}
+			if (NVS_ID_GT(entry_sector_id, sector_hdr.fd_id)) {
+				entry_sector_id = sector_hdr.fd_id;
+				fs->entry_sector = i;
+			}
+			if (NVS_ID_GT(sector_hdr.fd_id, fs->sector_id)) {
+				fs->sector_id = sector_hdr.fd_id;
+			}
+		}
+	}
+
+	if ((invalid_sectors) || (!sectors_in_use) || (fs->locked)) {
+		/* invalid sectors found or only empty sectors or reinit of
+		 * locked fs -> clear the fs
+		 */
+		SYS_LOG_DBG("(Re)Initializing sectors");
+		rc = nvs_clear(fs);
+		if (rc) {
+			goto err;
+		}
+		rc = _nvs_sector_init(fs, 0);
+		if (rc) {
+			goto err;
+		}
+		fs->entry_sector = 0;
+		fs->locked = false;
 
-	rc = nvs_get_first_entry(fs, entry);
-	if (rc) {
-		return rc;
+		rc = 0;
+		goto err;
 	}
-	latest.id = entry->id;
-	latest.data_addr = entry->data_addr;
-	latest.len = entry->len;
+
+	/* Valid sectors found in flash, need to set fs->write_location */
+	addr = fs->entry_sector * fs->sector_size;
 	while (1) {
-		hdr_addr = _nvs_head_addr_in_flash(fs, entry);
-		rc = nvs_flash_read(fs, hdr_addr, &head, sizeof(head));
+		rc = _nvs_flash_rd_len_id_start(fs, addr, &id, &len1);
 		if (rc) {
-			return rc;
+			goto err;
 		}
-		if (head.id == NVS_ID_EMPTY) {
-			entry->id = latest.id;
-			entry->data_addr = latest.data_addr;
-			entry->len = latest.len;
-			return 0;
+		if (id == NVS_ID_EMPTY) {
+			break;
 		}
-		if (head.id == latest.id) {
-			latest.len = head.len;
-			latest.data_addr = entry->data_addr;
+		entry_len = get_wbs_aligned_entry_len(fs, len1);
+		_nvs_addr_advance(fs, &addr, entry_len);
+		rc = _nvs_flash_rd_len_id_end(fs, addr, &id, &len2);
+		if (len1 != len2) {
```

Comment:
I think we still need an automatic mechanism for restore FS automatically after the power down while writing issue - it is something we really need for any device in the field. this comment is continuation of the https://github.com/zephyrproject-rtos/zephyr/pull/8141#discussion_r194684068 thread
cc @lemrey You opinion?

Reply:
It is definitely desirable, the scenario is not uncommon.
I will comment on #8367.